### PR TITLE
Feature/ts rewrite

### PR DESCRIPTION
**Pull request by Saegrov**

### CODE REVIEW

File: src/index.spec.ts [hunk 1]
```diff
@@ -28,17 +18,17 @@ describe('isEqual', function () {
     expect(isEqual(null, undefined)).to.equal(true);
     expect(isEqual(undefined, undefined)).to.equal(true);
     expect(isEqual(NaN, NaN)).to.equal(true);
-    expect(isEqual(null, NaN)).to.equal(true);
-    expect(isEqual(undefined, NaN)).to.equal(true);
+    expect(isEqual(null, NaN)).to.equal(false);
```
Comment (esseb) [inline src/index.spec.ts]: Why was this changed?

If intentional, the test should be moved out of the test block that says it tests for `true`.

File: src/index.spec.ts [hunk 2]
```diff
@@ -28,17 +18,17 @@ describe('isEqual', function () {
...
+    expect(isEqual(undefined, NaN)).to.equal(false);
```
Comment (esseb) [inline src/index.spec.ts]: Why was this changed?

If intentional, the test should be moved out of the test block that says it tests for `true`.

File: src/index.spec.ts [hunk 3]
```diff
@@ -28,17 +18,17 @@ describe('isEqual', function () {
...
   });
   it('should return "true" for objects with same properties', function () {
     expect(isEqual({ foo: 'foo', bar: true }, { bar: true, foo: 'foo' })).to.equal(true);
     expect(isEqual({ foo: null }, { foo: null })).to.equal(true);
     expect(isEqual({ foo: null }, { foo: undefined })).to.equal(true);
     expect(isEqual({ foo: undefined }, { foo: undefined })).to.equal(true);
     expect(isEqual({ foo: NaN }, { foo: NaN })).to.equal(true);
-    expect(isEqual({ foo: null }, { foo: NaN })).to.equal(true);
-    expect(isEqual({ foo: undefined }, { foo: NaN })).to.equal(true);
+    expect(isEqual({ foo: null }, { foo: NaN })).to.equal(false);
```
Comment (esseb) [inline src/index.spec.ts]: Why was this changed?

If intentional, the test should be moved out of the test block that says it tests for `true`.

File: src/index.spec.ts [hunk 4]
```diff
@@ -28,17 +18,17 @@ describe('isEqual', function () {
...
+    expect(isEqual({ foo: undefined }, { foo: NaN })).to.equal(false);
```
Comment (esseb) [inline src/index.spec.ts]: Why was this changed?

If intentional, the test should be moved out of the test block that says it tests for `true`.

File: src/index.ts [hunk 5]
```diff
@@ -1,28 +1,22 @@
...
- * @param {Object} obj2
- * @param {Array} [ignore]
- * @param {Debug} [debug]
- * @returns {Boolean}
+ * Determine if 'obj1' and 'obj2' are conceptually equal optionally ignoring properties in 'ignoredObjectProps'
  */
-module.exports = function isEqual (obj1, obj2, ignore, debug) {
-  ignore = ignore || [];
+export function isEqual(obj1: any, obj2: any, ignoredObjectProps?: Array<string>, debug?: DebugFn): boolean {
+  ignoredObjectProps = ignoredObjectProps || [];
```
Comment (esseb) [inline src/index.ts]: Can we use default function parameter for `ignoredObjectProps` instead of this line? Not sure how this looks with typescript.

File: src/index.ts [hunk 6]
```diff
@@ -39,21 +33,17 @@ module.exports = function isEqual (obj1, obj2, ignore, debug) {
...
-function equal (val1, val2) {
-  const type1 = typeof val1;
-  const type2 = typeof val2;
+function equal(val1: any, val2: any): boolean {
 
-  // Convert NaN to null
-  if (type1 == 'number' && isNaN(val1)) val1 = null;
-  if (type2 == 'number' && isNaN(val2)) val2 = null;
+  // Special handling for NaN
+  if (typeof val1 == 'number' && isNaN(val1) && typeof val2 == 'number' && isNaN(val2)) {
```
Comment (esseb) [inline src/index.ts]: I think merging this into one test changes the logic a bit when e.g. `val1` is `NaN` and `val2` is `null`, which is why some testcases changed from `true` to `false`.

File: src/index.ts [hunk 7]
```diff
@@ -63,38 +53,29 @@ function equal (val1, val2) {
...
  * Determine if 'obj' is an object
- * @param {Object} obj
- * @returns {Boolean}
  */
-function isObject (obj) {
+function isObject(obj: any): boolean {
   const type = typeof obj;
 
   return 'object' == type
-    && 'function' != type
```
Comment (esseb) [inline src/index.ts]: Why was this line removed?

File: src/index.ts [hunk 8]
```diff
@@ -4,90 +4,109 @@
...
-  if (isObject(obj1) && isObject(obj2)) {
-    const keys1 = keys(obj1, ignoredObjectProps);
-    const keys2 = keys(obj2, ignoredObjectProps);
+function isEqualObject(obj1: Indexable, obj2: Indexable, ignoredObjectProps: string[], debug?: DebugFn) {
+  const ignoredProps = stringArrayToObject(ignoredObjectProps);
+  const keys1 = nonFunctionKeys(obj1);
+  const keys2 = nonFunctionKeys(obj2);
 
-    if (keys1.length != keys2.length) return false;
+  if (keys1.length != keys2.length) {
```
Comment (esseb) [inline src/index.ts]: This logic changes now that you no longer filter out `ignoredObjectProps`.

If `ignoredObjectProps` is `['foo', 'bar']` `obj1` has `{ name: 'Name', foo: 'foo' }` and `obj2` has `{ name: 'Name', foo: 'foo', bar: 'bar' }` they would be equal before because both would have length 1 after stripping out the ignored object properties, but since you don't strip ignored object properties any more they will now be non-equal since they have length 2 and 3 respectively.

Comment (esseb) [reply src/index.ts]: I guess there isn't a testcase for this case, we should add one.

**Review by esseb — COMMENTED** [review]
I think the PR does too much. Changes logic and readability when it should just be about converting to typescript.

We can change behavior, but we need to do so for a reason, and in a separate PR.

File: src/index.ts [hunk 9]
```diff
@@ -4,90 +4,109 @@
...
-  });
+function nonFunctionKeys(obj: Indexable) {
+  return Object.keys(obj).filter(key => typeof obj[key] !== 'function');
 }
 
 /**
  * Determine if arrays 'arr1' and 'arr2' are equal
  */
-function isEqualArray(arr1: any, arr2: any): boolean {
-  if (Array.isArray(arr1) && Array.isArray(arr2)) {
```
Comment (esseb) [inline src/index.ts]: Removing this line means a `['foo']` will equal `{'0': 'foo', length: 1}` I think.

File: src/index.ts (hunk 8 above)
Comment (Saegrov) [reply src/index.ts]: Good catch! Will add test in another PR.

**Review by esseb — APPROVED** [review]
